src/infer.ts

```diff
@@ -12,28 +12,138 @@ import ts, { FunctionDeclaration, StringLiteralLike } from "npm:typescript@5.1.6
+  name: string
+  kind: "scalar" | "object"
+}
+
+export enum NullOrUndefinability {
+  AcceptsNullOnly = "AcceptsNullOnly",
+  AcceptsUndefinedOnly = "AcceptsUndefinedOnly",
+  AcceptsEither = "AcceptsEither",
+}
+
+export function get_ndc_schema(programInfo: ProgramSchema): sdk.SchemaResponse {
+  const functions = Object.entries(programInfo.functions);
+
+  const object_types = mapObjectValues(programInfo.object_types, obj_def => {
+    return {
+      fields: Object.fromEntries(obj_def.properties.map(prop_def => [prop_def.property_name, { type: convert_type_definition_to_sdk_type(prop_def.type)}]))
+    }
+  });
 
-export type ProgramInfo = {
-  schema: sdk.SchemaResponse,
-  positions: FunctionPositions
+  const scalar_types = mapObjectValues(programInfo.scalar_types, _scalar_def => {
+    return {
+      aggregate_functions: {},
+      comparison_operators: {},
+    }
+  })
+
+  return {
+    functions: functions
+      .filter(([_, def]) => def.ndc_kind === FunctionNdcKind.Function)
+      .map(([name, def]) => convert_function_definition_to_sdk_schema_type(name, def)),
+    procedures: functions
+      .filter(([_, def]) => def.ndc_kind === FunctionNdcKind.Procedure)
+      .map(([name, def]) => convert_function_definition_to_sdk_schema_type(name, def)),
+    collections: [],
+    object_types,
+    scalar_types,
+  }
 }
 
-function mapObject<I, O>(obj: {[key: string]: I}, fn: ((key: string, value: I) => [string, O])): {[key: string]: O} {
-  const keys: Array<string> = Object.keys(obj);
-  const result: {[key: string]: O} = {};
-  for(const k of keys) {
-    const [k2, v] = fn(k, obj[k]);
-    result[k2] = v;
+function convert_type_definition_to_sdk_type(typeDef: TypeDefinition): sdk.Type {
+  switch (typeDef.type) {
+    case "array": return { type: "array", element_type: convert_type_definition_to_sdk_type(typeDef.element_type) }
+    case "nullable": return { type: "nullable", underlying_type: convert_type_definition_to_sdk_type(typeDef.underlying_type) }
+    case "named": return { type: "named", name: typeDef.name }
+    default: return unreachable(typeDef["type"])
   }
-  return result;
 }
 
-function isParameterDeclaration(node: ts.Node | undefined): node is ts.ParameterDeclaration {
-  return node?.kind === ts.SyntaxKind.Parameter;
+function convert_function_definition_to_sdk_schema_type(function_name: string, definition: FunctionDefinition): sdk.FunctionInfo | sdk.ProcedureInfo {
+  const args =
+    definition.arguments
+      .map(arg_def =>
+        [ arg_def.argument_name,
+          {
+            type: convert_type_definition_to_sdk_type(arg_def.type),
+            ...(arg_def.description ? { description: arg_def.description } : {}),
+          }
+        ]
+      );
+
+  return {
+    name: function_name,
+    arguments: Object.fromEntries(args),
+    result_type: convert_type_definition_to_sdk_type(definition.result_type),
+    ...(definition.description ? { description: definition.description } : {}),
+  }
```

Comment:
We do the conversion every time you call the schema endpoint. It's annoying, but we can't avoid doing that unless we put it inside the configuration, which seems like an unnecessary duplication in memory. The schema endpoint isn't called very much, only when you refresh the Data Connector using the LSP, so that seems like a better trade off than a 2x size Configuration IMHO.

Reply:
Yeah I guess it's ok on /schema since that's only called during "dev" and this simplifies things.